### DISCUSSION TITLE
Hide unused navigation bar options by user permissions

### DISCUSSION THREAD

**Discussion by VazquezFrancisco** [netbox-community/netbox#8706] [Q&A]
Hi,
I need to hide the unused options in the left navigation bar.
I managed to hide them statically just commenting lines at the MENUS section starting from line 372 in the /opt/netbox/netbox/netbox/navigation_menu.py file.
![imagen](https://user-images.githubusercontent.com/98488241/155004099-488229a1-3df0-4155-a14b-c52fd1ca3f8d.png)
![imagen](https://user-images.githubusercontent.com/98488241/155004130-386b9681-1900-4edf-ada8-7af6c5838dbe.png)

The goal now is to hide the unused options depending by the users permissions. If an user doesn't have view permission to an specific option, then he shouldn't see it (now the user can see it with a lock icon). It can be done somehow?

**Comment by candlerb** [comment]
I personally agree with you: I would prefer that any area that the user doesn't have access to should be hidden.  Implementation-wise this would be trivial: instead of displaying a pane with a lock icon inside it, just don't display the pane - and hide the corresponding menu item.

Jeremy said recently (#7800) that this would confuse users who can't see things that the documentation says should be present. But I note this already happens in some areas: for example you don't see a "Device Bays" tab on a device if it has no bays, you don't see the napalm tabs if the device Platform doesn't allow it, and you don't see the "Admin" menu item if you don't have the "is_staff" privilege.  Furthermore, a local Netbox admin with "is_superuser" privilege will be able to see everything - and therefore can easily diagnose that if one of their local end-users can't see something, it's because their permissions don't allow it.

There is an original issue at #426, which pre-dates the permissions capability in Netbox.

**Comment by VazquezFrancisco** [comment]
I've managed it with the help of Jason Yates from Netbox NetDev community. He edited the /opt/netbox/netbox/utilities/templates/navigation/menu.html code to this:

```
{% load helpers %}

<ul class="navbar-nav">
    {% for menu in nav_items %}
        <li class="nav-item">
            <a class="nav-link" href="#menu{{ menu.label }}" data-bs-toggle="collapse" role="button" aria-expanded="false" aria-controls="menu{{ menu.label }}" id="menu{{ menu.label }}" style="display:none">
               <i class="{{ menu.icon_class }}"></i>
               <span class="nav-link-text">{{ menu.label }}</span>
             </a>
            <div class="collapse" id="menu{{ menu.label }}">
                <ul class="nav nav-sm flex-column">

                    {% for group in menu.groups %}
                        {# Within each main menu, there are groups of menu items #}
                            <li class="nav-item">
                                {# Group Label #}
                                <div class="nav-group-header" id="group{{ group.label }}" style="display:none">
                                    <span class="nav-group-label">{{ group.label }}</span>
                                </div>
                            </li>

                        {% for item in group.items %}
                            {# Each Item #}
                            {% if request.user|has_perms:item.permissions %}
                            <li class="nav-item{% if not item.buttons %} no-buttons{% endif %}">
                                <a href="{% url item.link %}" class="nav-link">
                                    {{ item.link_text }}
                                </a>

                                {# Menu item buttons (if any) #}
                                {% if item.buttons %}
                                    <div class="btn-group px-2">
                                        {% for button in item.buttons %}
                                            {% if request.user|has_perms:button.permissions %}
                                                    <a class="btn btn-sm btn-{{ button.color }} lh-1" href="{% url button.link %}" title="{{ button.title }}">
                                                        <i class="{{ button.icon_class }}"></i>
                                                    </a>
                                            {% endif %}
                                        {% endfor %}
                                    </div>
                                {% endif %}
                            </li>
                            {% endif %}
                        {% endfor %}
                    {% endfor %}
                </ul>
            </div>
        </li>
    {% endfor %}
</ul>
<script type="application/javascript">
{% for menu in nav_items %}
    {% for group in menu.groups %}
        {% for item in group.items %}
            {% if request.user|has_perms:item.permissions %}
                document.getElementById("menu{{ menu.label }}").style.display = "block";
                document.getElementById("group{{ group.label }}").style.display = "block";
            {% endif %}
        {% endfor %}
    {% endfor %}
{% endfor %}
</script>
```

Now I see the options in the navigation menu depending of the menu items permission:

![imagen](https://user-images.githubusercontent.com/98488241/155135983-2135329e-1fea-4e2b-820e-a5c08e15b43b.png)

Also I've made some other changes in /opt/netbox/netbox/netbox/navigation_menu.py to hide unused navigation submenu options, and in /opt/netbox/netbox/templates/home.html to empty the unused boxes at the main screen.

Thanks for you response.

**Reply by qwertycandy** [reply]
Hello, NetBox newb here :) Thank you for the solution, I just have a question - is it necesarry to restart NetBox for the changes to take effect, or do something similar? I've first tested this in my Docker environment and everything worked great, but when I edit the menu.html in NetBox that's running on a server, nothing seems to change. Thank you.

**Reply by VazquezFrancisco** [reply]
> Hello, NetBox newb here :) Thank you for the solution, I just have a question - is it necesarry to restart NetBox for the changes to take effect, or do something similar? I've first tested this in my Docker environment and everything worked great, but when I edit the menu.html in NetBox that's running on a server, nothing seems to change. Thank you.

You're welcome! Yes you need to restart.

Also, there's an issue to add this to a newer version of Netbox:
https://github.com/netbox-community/netbox/issues/9070#

**Reply by qwertycandy** [reply]
Thank you very much, especially for replying so quickly - your solution helped me a lot.

**Comment by pboers1988** [comment]
@VazquezFrancisco I can confirm it works? We would be interested with this a feature? Are you going to make a PR?

**Reply by VazquezFrancisco** [reply]
> @VazquezFrancisco I can confirm it works? We would be interested with this a feature? Are you going to make a PR?

It works perfect, it would be great to have it as a feature by default.
I'm not going to make a PR, but anyone is allowed to use this and do anything with it.

Related files attached.
[Files.zip](https://github.com/netbox-community/netbox/files/8967117/Files.zip)

**Reply by emersonfelipesp** [reply]
There is already a issue about this. Check below:
**[Toggle the visibility of navigation menu items](https://github.com/netbox-community/netbox/issues/9070#)**
